### PR TITLE
[fix][client] Fix subscription topic name error.

### PR DESCRIPTION
**Pull request by shibd**
#16374  #16556 #16375 #16599

### Motivation

This is a bug, after [[PIP-145](https://github.com/apache/pulsar/pull/16062)],  `PatternMultiTopicsConsumer` can receive  `CommandWatchTopicUpdate` from broker to subscribe new topic.

But this topic name is with the partition index(case: public/default/test-topic-partition-0), This will produce unexpected behavior when executing the following subscribe method.

https://github.com/apache/pulsar/blob/9c93ab45af80dbeb116bfe9b63ff579ac4e22ce6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L935-L942

The reasons for these failed unit tests are:

When the client receives the `CommandWatchTopicUpdate` command, it will enter the logic of processing non-partitions, then only process the [consumer](https://github.com/apache/pulsar/blob/9c93ab45af80dbeb116bfe9b63ff579ac4e22ce6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L79) collection and not the [partitionedTopics](https://github.com/apache/pulsar/blob/9c93ab45af80dbeb116bfe9b63ff579ac4e22ce6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L82) collection

https://github.com/apache/pulsar/blob/9c93ab45af80dbeb116bfe9b63ff579ac4e22ce6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L1068-L1094

Then, when we execute the `consumer1.run(consumer1.getRecheckPatternTimeout())` to trigger the update subscription on the unit test, By executing the following logic, will cause `topic-4` to be removed and added at the same time. ultimately destabilizes unit tests (maybe deletes will be executed after additions)

https://github.com/apache/pulsar/blob/9c93ab45af80dbeb116bfe9b63ff579ac4e22ce6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java#L115-L121

Print log here:
image

Can See:

```
2022-07-21T19:43:35,883 - INFO  - [pulsar-client-io-397-1:PatternMultiTopicsConsumerImpl@154] - debug add: [persistent://my-property/my-ns/pattern-topic-4-AutoSubscribePatternConsumer]
2022-07-21T19:43:35,883 - INFO  - [pulsar-client-io-397-1:PatternMultiTopicsConsumerImpl@155] - debug remove: [persistent://my-property/my-ns/pattern-topic-4-AutoSubscribePatternConsumer-partition-0, persistent://my-property/my-ns/pattern-topic-4-AutoSubscribePatternConsumer-partition-1, persistent://my-property/my-ns/pattern-topic-4-AutoSubscribePatternConsumer-partition-2, persistent://my-property/my-ns/pattern-topic-4-AutoSubscribePatternConsumer-partition-3]

```

### Modifications

- Change subscribe logic, use the `getPartitionedTopicName()`.

### Documentation

- [x] `doc-not-needed` 
(Please explain why)

### CODE REVIEW

**Comment by shibd** [comment]
@codelipenghui @gaozhangmin @Technoboy- @nodece Please help review.

**Review by codelipenghui — APPROVED** [review]
Nice catch!

**Review by Technoboy- — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]
Good catch!

**Comment by mattisonchao** [comment]
@andrasbeni Could you help to review it?

**Review by nodece — APPROVED** [review]

**Review by BewareMyPower — CHANGES_REQUESTED** [review]
Could you add a unit test to protect this change?

**Comment by shibd** [comment]
> Could you add a unit test to protect this change?

@BewareMyPower Thanks for your reminder, I looked again, and It is found that we cannot be modify the logic in the `MultiTopicsConsumerImpl.subscribeAsync` method. Because users may use consumer like this:

``` java
        client.newConsumer().topic(
                        "persistent://public/dafault/test-topic1-partition-0",
                        "persistent://public/dafault/test-topic2-partition-1"
         );
```
If force subscription to PartitionTopic inside the method, Then all partitions of `test-topic1` and `test-topic2` will be subscribed. This is a breaking change.

So, I revert the original changes.

The new changes are: Inside the `PatternTopicsChangedListener.onTopicsAdded` method let it use partitioned name to subscribe.

@codelipenghui @mattisonchao @Technoboy- @nodece @BewareMyPower Sorry, please help review again.

**Comment by codelipenghui** [comment]
> The new changes are: Inside the PatternTopicsChangedListener.onTopicsAdded method let it use partitioned name to subscribe.

Make sense.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java [hunk 1]
```diff
@@ -514,6 +509,38 @@ public void testStartEmptyPatternConsumer() throws Exception {
         producer3.close();
     }
 
+    @Test(timeOut = testTimeout)
+    public void testAutoSubscribePatterConsumerFromBrokerWatcher() throws Exception {
+        String key = "AutoSubscribePatternConsumer";
+        String subscriptionName = "my-ex-subscription-" + key;
+
+        Pattern pattern = Pattern.compile("persistent://my-property/my-ns/pattern-topic.*");
+        Consumer<byte[]> consumer = pulsarClient.newConsumer()
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:518]: Please close the consumer after the test.

**Review by codelipenghui — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java [hunk 2]
```diff
@@ -203,9 +205,14 @@ public CompletableFuture<Void> onTopicsAdded(Collection<String> addedTopics) {
                 return addFuture;
             }
 
+            Set<String> addTopicPartitionedName = addedTopics.stream()
```
Comment (nodece) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java:208]: Do you need to fix the https://github.com/apache/pulsar/pull/16719/files#diff-779a47f80b0048bab183870ed4c1ecd07ffd538006cf9358e494bee23405d924R179?

There has the same logic with this.

Comment (shibd) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java:208]: Do you mean removeTopic? there is no problem with the implementation of removeTopic.

Comment (nodece) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java:208]: Okay.

**Review by nodece — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java [hunk 3]
```diff
@@ -203,9 +205,14 @@ public CompletableFuture<Void> onTopicsAdded(Collection<String> addedTopics) {
...
+                    .map(addTopicName -> TopicName.get(addTopicName).getPartitionedTopicName())
+                    .collect(Collectors.toSet());
+
             List<CompletableFuture<Void>> futures = Lists.newArrayListWithExpectedSize(partitionedTopics.size());
-            addedTopics.stream().forEach(topic -> futures.add(
-                    subscribeAsync(topic, false /* createTopicIfDoesNotExist */)));
+            addTopicPartitionedName.stream().forEach(partitionedTopic -> futures.add(
```
Comment (BewareMyPower) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PatternMultiTopicsConsumerImpl.java:213]:
```suggestion
            addTopicPartitionedName.forEach(partitionedTopic -> futures.add(
```

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java [hunk 4]
```diff
@@ -469,11 +469,6 @@ public void testStartEmptyPatternConsumer() throws Exception {
             .messageRoutingMode(org.apache.pulsar.client.api.MessageRoutingMode.RoundRobinPartition)
             .create();
 
-        List<String> topicNames = Lists.newArrayList(topicName1, topicName2, topicName3);
-        NamespaceService nss = pulsar.getNamespaceService();
-        doReturn(CompletableFuture.completedFuture(topicNames)).when(nss)
-                .getListOfPersistentTopics(NamespaceName.get("my-property/my-ns"));
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:475]: Why did you remove these three lines?

Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PatternTopicsConsumerImplTest.java:475]: This is a PR added before trying to fix that test, it didn't work.

https://github.com/apache/pulsar/pull/16599/files

**Review by BewareMyPower — APPROVED** [review]
